aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,158 @@
+
+The second test verifies that when ``ModelState`` is valid, a new ``BrainStormSession`` is added (via the repository), and the method returns a ``RedirectToActionResult`` with the expected properties. Mocked calls that aren't called are normally ignored, but calling ``Verifiable`` at the end of the setup call allows it to be verified in the test. This is done with the call to ``mockRepo.Verify``.
+
+Another controller in the app displays information related to a particular brainstorming session. It includes some logic to deal with invalid id values:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Controllers/SessionController.cs
+  :language: c#
+  :emphasize-lines: 16,20,25,33
+
+Looking at this controller action, it should be clear that there are at least three cases to test, one for each ``return`` statement. The unit tests are shown below:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/SessionControllerIndex.cs
+  :language: c#
+  :emphasize-lines: 16,26,39
+
+Finally, the application exposes some functionality as a web API, including a list of ideas associated with a brainstorming session and a method for adding new ideas to a session. The controller is shown below:
+
+.. _ideas-controller:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Api/IdeasController.cs
+  :language: c#
+  :emphasize-lines: 20-22,27,29-35,49-51,55,60,70
+
+The ``ForSession`` method returns a list of ``dynamic`` types, with property names camel cased to match JavaScript conventions. Avoid returning your business domain entities directly via API calls, since frequently they include more data than the API client requires, and they unnecessarily couple your app's internal domain model with the API you expose externally. You can define strongly typed data-transfer objects (DTOs), or just return dynamic as this method does. Mapping between domain entities and the types you will return over the wire can be done manually (using a LINQ ``Select`` as shown here) or using a library like `AutoMapper <https://github.com/AutoMapper/AutoMapper>`_
+
+.. note:: Returning ``dynamic`` types is a common practice, especially in web API methods, since it's a simple, easy way to control the shape of the output. However, ``dynamic`` types are ``internal``, so unit tests that attempt to refer to their properties may encounter errors unless the web project is configured to make its internals visible to the test project.
```

Comment:
In my experience it is a quick and easy step that gets the code away from returning the underlying model (which is a bad practice) and toward returning a custom DTO for the API to use. If the type is used in only one place, then just returning .Select(s => new { s.Id, s.Name }) is much easier and less verbose than having to define a DTO type.
Reference: http://stackoverflow.com/questions/10123371/returning-anonymous-types-with-web-api
It's also a very simple way to perform mapping to the client model format (e.g. changing casing to follow javascript conventions) without having to mess with the global formatter settings, etc.